pontoon/sync/formats/xliff.py

```diff
@@ -88,33 +68,123 @@ def sync_changes(self):
+                raise ParseError(f"Failed to parse {path}: {err}")
+
+            # Loop through each unit in the XLIFF file
+            for order, unit in enumerate(self.xliff_file.units):
+                # Get the unit's ID and source string
+                key = unit.getid()
+                context = unit.xmlelement.get("id")
+                source_string = unit.rich_source[0]
+                source_string_plural = ""
+
+                # Get the translated string for the unit. If there's no target string, this will be an empty dictionary
+                target_string = unit.get_rich_target()[0] if unit.get_rich_target() else None
+                strings = {None: target_string} if target_string else {}
+
+
+                # Get the unit's comments, split by newline characters
+                comments = unit.getnotes().split("\n") if unit.getnotes() else []
+
+                # Create a new XLIFFEntity from the unit
+                entity = XLIFFEntity(
+                    key,
+                    context,
+                    source_string,
+                    source_string_plural,
+                    strings,
+                    comments,
+                    order,
+                    unit
+                )
+                print(f"Adding entity with key {entity.key} to entities")
+                # Add the entity to the entities dictionary using its key as the dictionary key
+                self.entities[entity.key] = entity
 
     @property
     def translations(self):
-        return self.entities
+        return sorted(self.entities.values(), key=lambda e: e.order)
 
     def save(self, locale):
-        for entity in self.entities:
-            entity.sync_changes()
-
-        locale_code = locale.code
-
-        # TODO: should be made iOS specific
-        # Map locale codes for iOS: http://www.ibabbleon.com/iOS-Language-Codes-ISO-639.html
+        if not self.source_resource:
+            raise SyncError(
+                "Cannot save XLIFF resource {}: No source resource given.".format(
+                    self.path
+                )
+            )
+        # Open the file at the provided path
+        with open(self.source_resource.path, "r") as f:
+            # Read the contents of the file and encode it 
+            xml = f.read().encode("utf-8")
+
+            try:
+                # Parse the xml content of the file into an XLIFF file object
+                self.xliff_file = xliff.xlifffile(xml)
+            except etree.XMLSyntaxError as err:
+                # If there is an error parsing the file, raise a ParseError 
+                raise ParseError(f"Failed to parse {self.source_resource.path}: {err}")
+
+            # Loop through each unit in the XLIFF file
+            for order, unit in enumerate(self.xliff_file.units):
+                 # Apply any changes made to this object to the backing unit in the xliff file.
+                if None in unit.strings:
```

Comment:
In `sync_changes()`, `self` refers to the `XLIFFEntity` instance, not `unit`. You need to modify this line to `if None in entity.strings:` and get `entity` from `self.entities`. See [ftl.py](https://github.com/mozilla/pontoon/blob/2ad1cd1e987073bc0d175a75d9b1299550e3afa1/pontoon/sync/formats/ftl.py#L143-L144) for reference. You can get the key from the unit.